daemon/sailjailclient.c

```diff
@@ -0,0 +1,1284 @@
+    g_free(args);
+
+    if( !booster_name )
+        client_notify_launch_canceled(self, desktop_path);
+
+EXIT:
+    g_free(desktop_name);
+    g_free(binary_path);
+    g_free(booster_path);
+
+    return EXIT_FAILURE;
+}
+
+/* ------------------------------------------------------------------------- *
+ * CLIENT_NOTIFY
+ * ------------------------------------------------------------------------- */
+
+static void
+client_notify_launch_status(client_t *self, const char *method,
+                            const char *desktop)
+{
+    GDBusConnection *connection = client_session_bus(self);
+    if( connection ) {
+        /* We do not care too much about whether the call succeeds or not */
+        g_dbus_connection_call(connection, LAUNCHNOTIFY_SERVICE,
+                               LAUNCHNOTIFY_OBJECT, LAUNCHNOTIFY_INTERFACE,
+                               method, g_variant_new("(s)", desktop),
+                               NULL, G_DBUS_CALL_FLAGS_NO_AUTO_START,
+                               -1, NULL, NULL, NULL);
+        /* But we do want it to go out before we exec*() / exit() */
+        g_dbus_connection_flush_sync(connection, NULL, NULL);
+    }
+}
+
+static void
+client_notify_launching(client_t *self, const char *desktop)
+{
+    client_notify_launch_status(self, LAUNCHNOTIFY_METHOD_LAUNCHING, desktop);
+}
+
+static void
+client_notify_launch_canceled(client_t *self, const char *desktop)
+{
+    client_notify_launch_status(self,LAUNCHNOTIFY_METHOD_LAUNCH_CANCELED,
+                                desktop);
+}
+
+/* ========================================================================= *
+ * SAILJAILCLIENT
+ * ========================================================================= */
+
+static int
+sailjailclient_get_field_code(const char *arg)
+{
+    // Non-null string starting with a '%' followed by exactly one character
+    return arg && arg[0] == '%' && arg[1] && !arg[2] ? arg[1] : 0;
+}
+
+static bool
+sailjailclient_is_option(const char *arg)
+{
+    // Non-null string starting with a hyphen
+    return arg && arg[0] == '-';
+}
+
+static bool
+sailjailclient_match_argv(const char **tpl_argv, const char **app_argv)
+{
+    bool matching = false;
+
+    /* Rule out template starting with a field code */
+    if( sailjailclient_get_field_code(*tpl_argv) ) {
+        log_err("Exec line starts with field code");
+        goto EXIT;
+    }
+
+    /* Match each arg in template */
+    for( ;; ) {
+        const char *want = *tpl_argv++;
+
+        if( !want ) {
+            /* Template args exhausted */
+            if( *app_argv ) {
+                /* Excess application args */
+                log_err("argv has unwanted '%s'", *app_argv);
+                goto EXIT;
+            }
+            break;
+        }
+
+        int field_code = sailjailclient_get_field_code(want);
+
+        if( !field_code ) {
+            /* Exact match needed */
+            if( g_strcmp0(*app_argv, want) ) {
```

Comment:
Any reason for mixed use of `strcmp()` and `g_strcmp0()` ?

Reply:
Just that it is hard to write g_strcmp0() when you know that there are no NULLs involved. But then again some time one just might be wrong about it...

Reply:
Personally I'd just keep one of them in use but this was just a side note when browsing through the code.